Mathlib/Analysis/LocallyConvex/Separation.lean

```diff
@@ -312,4 +316,80 @@ theorem iInter_halfSpaces_eq (hs₁ : Convex ℝ s) (hs₂ : IsClosed s) :
   obtain ⟨l, s, hlA, hl⟩ := geometric_hahn_banach_closed_point (𝕜 := 𝕜) hs₁ hs₂ h
   obtain ⟨y, hy, hxy⟩ := hx l
   exact ((hxy.trans_lt (hlA y hy)).trans hl).false
+
+end
+
+section Countable
+
+variable [NormedAddCommGroup E] [NormedSpace ℝ E] [Module 𝕜 E] [ContinuousSMul 𝕜 E]
+
+/-- A closed convex set `s` is the intersection of countably many half spaces in a separable Banach
+space. Moreover, these halfspaces are all nontrivial if `s` is nonempty and not equal to `univ`. -/
+theorem iInter_nat_halfSpaces_eq
+    (hs₁ : Convex ℝ s) (hs₂ : IsClosed s) (hsep : IsSeparable sᶜ) :
+    ∃ (L : ℕ → E →L[𝕜] 𝕜) (c : ℕ → ℝ),
+    ⋂ (i : ℕ), {x | c i ≤ re (L i x)} = s ∧
+    (s.Nonempty → s ≠ univ → ∀ i, ∃ x, re (L i x) ≠ 0) := by
+  obtain rfl | hs₃ := s.eq_empty_or_nonempty
+  · exact ⟨0, 1, by simp [iInter_eq_empty_iff]⟩
+  obtain rfl | hs_univ := eq_or_ne s univ
```

Comment:
The proof of this theorem is rather long. It would be good if you document the proof with comments. It will make it easier for future readers to understand the proof, and optionally to refactor it.